Originated channel is now added to a bridge.

TACambron opened the pull request:
The program checks for an available bridge. If a bridge is available,
then that bridge is reused. If no bridges are active, a new bridge
is created. Once the bridge is created, a channel is originated to
a designated endpoint, and upon answering the call, the channel is
added to the bridge.

Commit:
Originated channel is now added to a bridge.

The program checks for an available bridge. If a bridge is available,
then that bridge is reused. If no bridges are active, a new bridge
is created. Once the bridge is created, a channel is originated to
a designated endpoint, and upon answering the call, the channel is
added to the bridge.

Commit:
Clarified naming, syntax and error handling.

Fixed many naming errors leftover from previous drafts. Program
now detects when the channel leaves the bridge and exits. This only
allows for one channel in the program, so later drafts will have to
adjust for this. Program also exits when a fatal error occurs. Also
fixed some sub-optimal syntax.

Commit:
Moved error message handling and fixed JSDoc syntax

Modified the original callback function to handle error messages
and added code to call the callback where errors can occur. Also
turned a multi-line comment into a JSDoc comment.

samuelg commented on app.js:
```diff
@@ -1,10 +1,27 @@
+/*jslint node: true */  
```
Look at adding a .jshintrc file to drive this type of jshint configuration. See https://raw.githubusercontent.com/asterisk/node-voicemail-data/master/.jshintrc for an example.

Update: You no longer need /*jslint node: true */ , your .jshintrc file is taking care of that.

samuelg replied:
You no longer need /*jslint node: true */ , your .jshintrc file is taking care of that.

samuelg replied:
/*jslint node: true */ is no longer necessary.

samuelg commented on app.js:
```diff
@@ -1,10 +1,27 @@
 var sla = require('./lib/sla.js');
 
-ari.connect('http://127.0.0.1:8088', 'ariUser', 'boogers', clientLoaded);
+ari.connect('http://127.0.0.1:8088', 'user', 'pass', clientLoaded);
```
You will eventually want to move the URL/user/pass of the connect method into a config file loaded by this file.

TACambron replied:
That is coming up in one of the next assignments.

samuelg commented on app.js:
```diff
@@ -1,10 +1,27 @@
+ * @param {Object} client - Object that contains information from the ARI 
+ *   connection.
+ */
+function clientLoaded (err, client) {
```
You never check to see if the connection returned an error.

Update: looks good.

samuelg commented on app.js:
```diff
@@ -1,10 +1,27 @@
+function clientLoaded (err, client) {
+  client.start('sla');
+  sla(client, function(err) {
+    if (err != null) {
```
You can just test for if (err)

samuelg replied:
BTW, if you needed to check against null, you should use !== instead of !=.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+ * @param {Object} client - Client received from app.js.
+ * @callback end - Handles errors and ends the application.
+ */
+module.exports = function(client, end) {
```
It's common to use the name callback for callback functions as it signals an asynchronous API.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+  /**
+   * Looks for an existing mixing bridge and creates one if none exist.
+   */
+  function designateMixingBridge() {
```
Because you are defining these functions inside the module.exports function, they will be created every time require is called on this file. You should move them above the export statement so they only need to be created once.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+      if (bridge) {
+        console.log('Using existing mixing bridge %s', bridge.id);
+      } else {
+        bridge = client.bridges.create({type: 'mixing'}, function(err, bridge) {
```
This is an asynchronous call, you can't assign the result of it to bridge. The bridge object that is passed into the callback is the object you are looking for.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+   */
+  function createChannel(client, bridge) {
+    var channel = client.Channel();
+    channel.originate({endpoint: 'SIP/phone', app: 'sla'},
```
The endpoint parameter should probably be passed into the module instead of being hard-coded.

wolfeman67 replied:
Configuration file will come later :)

TACambron replied:
Just for this assignment, Matt wanted us to hard code an endpoint. We will adjust this in later versions.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+        }
+      }
+    );
+    channel.on('StasisStart', function(event, outgoing) {
```
outgoing seems vague in this case. I would consider changing the name to something a bit more descriptive.

samuelg replied:
That particular channel is only ever going to enter into Stasis once. Use channel.once('StasisStart') instead of .on so the event handler can unregister itself once the event is raised and the handler can be garbage collected.

samuelg commented on package.json:
```diff
@@ -3,46 +3,28 @@
   "version": "1.0.0",
```
Read http://semver.org/. Typically, the first version of a module is 0.0.1. Once the first release is out, this changes to 0.1.0. After that, every time a feature is added, the middle number (minor version) is incremented. Every time a bug is fixed, the last number (patch version) is incremented. Every time a backwards incompatible change is added, the first number (major version) is incremented. Some modules are go to 1.0.0 once the API is finalized and the module is stable.

wolfeman67 replied:
Version number was fixed in one of the bridges, before being merged to master. An incorrect version number was used in master previously. But that has been corrected now.

samuelg commented on package.json:
```diff
@@ -3,46 +3,28 @@
   "version": "1.0.0",
   "description": "SLA Application using node-ari-client.js",
```
As far as NPM is concerned, the client is ari-client. node-ari-client is only the name on github.

samuelg commented on README.md:
```diff
@@ -1,2 +1,4 @@
 # node-ari-shared-line-appearances
 Co-op implementation of SLA in ARI using Node.js
+
+This implementation assumes that you have at least Asterisk 12.0.0 running. This is the version where ARI first came around, and it should have all of the functions required for this project. You must also have a valid ARI user in ari.conf named user and have a password "pass", as well as have 127.0.0.1 (or localhost) and 8088 configured as the bindaddr and bindport respectively in http.conf. You ...
```
Things like the ARI user/pass and the SIP endpoint (or other technologies such as PJSIP) should not be given here. These values should come from a configuration file.

samuelg commented on app.js:
```diff
@@ -1,10 +1,27 @@
+      process.exit(0);
+    }
+  });
+};
```
No need for a semi-colon here since you are using the function name() {} syntax.

samuelg commented:
Overall, instead of having a function that calls another function, which in turn calls another function and so on, consider having small functions responsible for a single thing that expect a callback so you can get the value you need (the bridge to add to, the channel object) and having the main export function call these in turn. This will make the module more readable and will help in maintenance down the line.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+        end(err);
+      }
+    });
+    bridge.on('ChannelEnteredBridge', function(event) {
```
That particular channel is only ever going to enter the bridge once. Use channel.once instead of .on so the event handler can unregister itself once the event is raised and the handler can be garbage collected.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+    bridge.on('ChannelEnteredBridge', function(event) {
+      console.log("Channel %s has entered the bridge", event.channel.id);
+    });
+    bridge.on('ChannelLeftBridge', function(event) {
```
That particular channel is only ever going to leave the bridge once. Use channel.once instead of .on so the event handler can unregister itself once the event is raised and the handler can be garbage collected.

Commit:
Fixed bridge creation and other syntax

Fixed a problem where creating a bridge caused the application to
fail due to bridge being referenced incorrectly for an asynchronous
program. Also fixed other small syntax issues, as well as moving
module.exports to the end of the file so that the functions will not
be created everytime that require is called on the file. Changed the
callback function's name to 'callback'.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+      }
+    }
+  );
+  channel.once('StasisStart', function(event) {
```
this should read: channel.once('StasisStart', function(event, channel) {

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+    }
+  );
+  channel.once('StasisStart', function(event) {
+    addChannelsToBridge(event.channel, bridge, callback);
```
use the channel object passed in as the second parameter for this. event.channel is a simple object literal that contains data only. The object passed in to the second parameter is an object with methods for further operations and responds to event listeners.

Commit:
Added parameters to event handling functions

Added parameters to catch information given off by events such as
for stasis start and channels entering or leaving a bridge. Also
removed the else statement in app.js since it was not necessary.

samuelg commented on app.js:
```diff
@@ -6,5 +7,23 @@ var sla = require('./lib/sla.js');
+  sla(client, function(err) {
+    if (err !== null) {
+      throw err;
+    } else {
```
No need for the else, throw will kill the process there.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
-'use strict';
+/*jslint node: true */
```
This should be removed now that you have a .jshintrc file.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
-    var str = "snarflegluff";
-    return str;}
+/**
+ * Looks for an existing mixing bridge and creates one if none exist.
```
jsdoc style param list is missing.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+function designateMixingBridge(client, callback) {
+  client.bridges.list(function(err, bridges) {
+    var bridge = bridges.filter(function(candidate) {
+      return candidate.bridge_type === 'mixing';
```
Is this the expected behavior? You never check to see if the bridge has channel(s) in it and since it's a mixing bridge, the new channel will hear whatever other channels are currently hearing and vice versa.

matt-jordan replied:
For this current task, yes, that's the expected behaviour. Later on, they'll have to be a bit smarter about how they select the bridges.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+    } else {
+      client.bridges.create({type: 'mixing'}, function(err, bridge) {
+        if (err) {
+          callback(err);
```
Just as an FYI, calling a callback does not exit the function at this point. Since you are not returning from the function, the call to createChannel will occur even though an error. This means that createChannel will eventually error out and call the callback yet again, which is very bad. This is another thing that is bad about callbacks; even though they should only be called once, it's very easy to accidentally call them multiple times.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+  bridge.addChannel({channel: channel.id}, function(err) {
+    if (err) {
+      channel.hangup();
+      callback(err);
```
See comment above, this will not actually exit the function and the event handlers will be registered even in the case of an error occurring.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,77 @@
+  channel.originate({endpoint: 'SIP/tcam', app: 'sla'},
+    function(err) {
+      if (err) {
+        callback(err);
```
See comment above, this will not actually exit the function and the event handler will be registered even in the case of an error occurring.

samuelg commented:
At this point, it would be a good idea to switch from callbacks to promises. You should use promises both internally to handle calling async APIs and also return a promise to users of your module that will eventually be resolved if everything works as expected, or rejected if an error occurs. You should take some time to read about promises and pick a library that implements them in Node.js.

Each of your method should have a single purpose, and should return a promise that can be used by the caller to determine if the module should be rejected, or if the next method should be called to handle the next part of the module's responsibilities.

In other words, the main function exposed as your module should look something like the following (note that I'm just reusing your current method names as an example, you may want to design the API differently):

return designateMixingBridge(args)
  .then(function(result) {
    // do one of these
    return createBridge(args);
    return createChannel(args);
  })
  .then(function(result) {
    return addChannelToBridge(args);
  });

Note that .done is not called on the promise since it is returned to the user of the module.

You will need to look into converting the Node.js style callback API of ari-client to a promise one using something like Q's denodeify method so you can call ari-client methods and get back promises.

samuelg commented:
Running jshint found the following errors:

lib/sla.js: line 2, col 1, Missing semicolon.
lib/sla.js: line 10, col 0, Identifier 'bridge_type' is not in camel case.
lib/sla.js: line 41, col 52, Mixed double and single quotes.
lib/sla.js: line 44, col 49, Mixed double and single quotes.
lib/sla.js: line 71, col 81, Line is too long.
lib/sla.js: line 72, col 81, Line is too long.
lib/sla.js: line 73, col 86, Line is too long.
lib/sla.js: line 75, col 86, Line is too long.

Commit:
Implemented promises

Replaced all callbacks to use promises from the Q promise library.
Much of the file has been restructured because of this. Module.exports
now contains nested then() statements instead of each function
calling the next.

samuelg commented on app.js:
```diff
@@ -1,10 +1,39 @@
+  .then(function (client) {
+    clientLoaded(client);
+  })
+  .catch(errHandler);
```
You should call done here so the error you throw in errHandler can propagate up the stack.

samuelg commented on app.js:
```diff
@@ -1,10 +1,39 @@
-function clientLoaded (err, client){
-    console.log(sla());}
+connect('http://127.0.0.1:8088', 'user', 'pass')
+  .then(function (client) {
```
you can just use .then(clientLoaded)

samuelg commented on app.js:
```diff
@@ -1,10 +1,39 @@
+    .then(console.log)
+    .catch(errHandler)
+    .finally(function () {
+      process.exit(0);
```
Since this is in a finally handler, process will exit with a non error 0 code even in the case of an error.

samuelg commented on app.js:
```diff
@@ -1,10 +1,39 @@
+  sla(client)
+    .then(console.log)
+    .catch(errHandler)
+    .finally(function () {
```
You should call done as the final call to the promise.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,93 @@
+ */
+function designateMixingBridge(client) {
+  var defer = Q.defer();
+  var list = Q.denodeify(client.bridges.list.bind(client.bridges));
```
You actually want to bind to client here, not client.bridges. Any time you denodeify a function called on client.someResource you pass the client to bind. Any time you denodeify a function called on a resource instance like bridge.get you pass the bridge to bind.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,93 @@
+          console.log('Using existing mixing bridge %s', bridge.id);
+          defer.resolve(bridge);
+        } else {
+          client.bridges.create({type: 'mixing'}, function(err, bridge) {
```
This is still using the callback version of bridges.create and should be switched to promise style.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,93 @@
+  });
+  bridge.once('ChannelLeftBridge', function(event, bridge) {
+    console.log('Channel %s has left the bridge', bridge.channel.id);
+    defer.resolve('Application completed');
```
That's fine for this first step, but I don't think you want to ultimately have the promise resolve only once the channel hangs up. Events are a better way of informing users of your API that something like that has happened.

samuelg commented on lib/sla.js:
```diff
@@ -1,6 +1,93 @@
+module.exports = function(client) {
+  return designateMixingBridge(client)
+    .then(function (bridge) {
+      return createChannel(client, bridge)
```
This is fine for this first step but there's probably a way in the future to make this flatter and not nest the calls like this.

Commit:
Resolved promise syntax errors

Fixed a few promise syntax errors such as removing the .finally()
statements and replacing them with .done() statements. Also fixed
client.bridges.create() inside of designateMixingBridge to be a
promise style rather than a callback style.

samuelg commented:
Looks good. Since you have a lot of commits to address PR comments, I would suggest we walk though squashing the commits into a single commit in a different branch before merging.

TACambron commented:
Squashed into #6
